### ISSUE TITLE
Refunds UX Improvements

### ISSUE THREAD

**Issue by roykho** [woothemes/woocommerce#7222]
- [x] Change the order status to refunded if the order is completely refunded.
- [ ] Update total_sales product postmeta when refund occurs https://github.com/woothemes/woocommerce/issues/7299
- [x] Refresh the order notes area after refund

---

So I've been working with refunds quite a bit as of late.  I feel we could improve on this for a better user experience for the admin.

I know that we had talks about putting the order status to "Refunded" if the order is completely refunded and keeping the status the same when it is a partial refund.

With that mind, currently our refund process is via AJAX.  But two things are less than ideal here.  One is that the order notes is not updated to show the refund process and the second issue would be that the order status is not updated.  Both of these items would need a full refresh to see.

Sure, you can use the AJAX callback to update these areas via JS but why not just then refresh the page and save some code.

The other thing is if a refund fails, it gives a very generic "Refund failed" alert to the admin.  I think we should allow payment gateways to pass in its own error message to the refund process so it is more descriptive on what is going on without having to look into the error log.

So to summarize, the two implementations I wish to add is refresh the page after a refund process and allow payment gateways to pass in its own refund errors "if any" to be alerted out.

I am happy to do PRs for this if you guys think it is a good idea.  @mikejolley @claudiosmweb

**Comment by michaeltieso** [comment]
To add to this it'd be great to be able to refund individual items and restock that product.
I also like the idea of adding it to the order notes.

**Comment by maxrice** [comment]
I just added refund/void support to our payment gateway framework and definitely agree with changing the order status to `refunded` if the order is completely refunded. We're already doing this in the framework so it'd be nice to have it as part of core instead :smile_cat: 

Also agree with refreshing the order notes area, it's a bit strange to render an error message in a dialog box and then force the user to fresh the page to see the actual note added (indeed it's not clear that a page refresh is required to see additional info).

I think the ability for gateways to pass back error messages already exists with the `WP_Error` return value, this is what it looks like for Authorize.net AIM:
![screen shot 2015-01-27 at 12 57 38 pm](https://cloud.githubusercontent.com/assets/1579862/5923852/957f1a0c-a624-11e4-834a-4a498b700edf.png)

Like @roykho said though, I think it'd be nicer to do this as part of a page refresh and and instead display a notice at the top of the page with the error?

One other thing that would be great is to pass the refund object to the `process_refund()` method somehow. Right now we're storing refund-related transaction data as order meta which works, but I think it'd be nicer to store them as refund "order" meta.

**Comment by mikejolley** [comment]
Added items to the OP :) @roykho good to go

**Comment by nishitlangaliya** [comment]
@mikejolley, Hi I have pushed the code for reload page after refunds and done code for display message with error code when refunds failed. ref #9b1b353

Please check for the same when you get a chance

Thanks :)

**Comment by mikejolley** [comment]
"Refresh the order notes area after refund" is a non issue - for partial refunds no notes are created. Full refunds refresh the page as intended, possibly fixed since this issue was created.

**Comment by mikejolley** [comment]
Looking at total_sales, it should be left as-is (storing the total sales regardless of future order status). The reason being, even if decreased during refund, it would require lots of extra logic to increase this should a refund be deleted or reversed. This will lead to issues keeping the value synced.

One possible solution would be to query and cache that value rather than keeping it stored in meta permanently, but I suspect this best left until we refactor reporting for performance. For now, won't fix.
